login: Improve phone number input

marhkb opened the pull request:
This is early WIP (code is still a mess). But who wants, can test it already.

https://user-images.githubusercontent.com/3630213/137650312-b9d02f64-f6fe-4f5f-80ee-7d374c2a70bf.mp4

fixes #117

melix99 commented:
I'm wondering if we can format the phone number while typing (maybe using [this crate](https://crates.io/crates/phonenumber)). That can be useful also for the user dialog.

melix99 commented:
> What is extremely fiddly is to set the caret position after each formatting

Yeah, maybe a better idea is formatting only when the entry is not in active.

BTW my intention with formatting the number is to leave some space between the country code and the phone number instead of using a different color. But if formatting doesn't work that great while editing I'd suggest to manually put a space between them and leaving the full formatting for when the entry is no more active.

marhkb committed:
login: Use a function to determine `use_test_dc`

marhkb committed:
login: Improve phone number input

This commit improves the phone number input during the login process.
The improvements include the following:

- Load the list of country infos from TDLib
- Add a specialized country info to that list in case the test data
  center is used
- Show a country selection combobox row above the phone number entry
- Update the calling code in the phone number entry when a country is
  selected from the combobox
- Update the selected country info in the combobox when the user changes
  the phone number within the entry
- Highlight the country calling code in the phone number entry
- Keep the highligted country code in the entry intact even when white-
  spaces are inserted
- Select the number except calling code in the phone number entry in
  case of an error
- Determine the system country code and initially set the corresponding
  calling code

The phone number is not formatted during input. The reason for this is
that this de facto only works with whitespaces. To realize this is very
complex and works in certain cases only insufficiently.

Even after the phone number has been confirmed with `Next`, the phone
number will not be formatted. This would make little sense: In some case
the phone number is correct is and the user doesn`t see it after confir-
mation anyway. In the other case, if the phone number is incorrect, it
cannot be formatted in most cases.

fixes #117

marhkb commented:
This is finally ready for review.

Quoting the main commit
```
This commit improves the phone number input during the login process.
The improvements include the following:

- Load the list of country infos from TDLib
- Add a specialized country info to that list in case the test data
  center is used
- Show a country selection combobox row above the phone number entry
- Update the calling code in the phone number entry when a country is
  selected from the combobox
- Update the selected country info in the combobox when the user changes
  the phone number within the entry
- Highlight the country calling code in the phone number entry
- Keep the highligted country code in the entry intact even when white-
  spaces are inserted
- Select the number except calling code in the phone number entry in
  case of an error
- Determine the system country code and initially set the corresponding
  calling code

The phone number is not formatted during input. The reason for this is
that this de facto only works with whitespaces. To realize this is very
complex and works in certain cases only insufficiently.

Even after the phone number has been confirmed with `Next`, the phone
number will not be formatted. This would make little sense: In some case
the phone number is correct is and the user doesn`t see it after confir-
mation anyway. In the other case, if the phone number is incorrect, it
cannot be formatted in most cases.

fixes https://github.com/melix99/telegrand/issues/117
```

melix99 commented on src/country_info.rs:
```diff
@@ -0,0 +1,171 @@
+
+#[derive(Clone, Debug, Default, glib::Boxed)]
+#[boxed_type(name = "BoxedCallingCodes")]
+pub(crate) struct BoxedCallingCodes(BTreeSet<String>);
```
```suggestion
pub(crate) struct CallingCodes(BTreeSet<String>);
```

It's ok calling it just CallingCodes in this case. I usually use the Boxed* prefix only when I need to wrap a boxed type around an existing type.

melix99 commented on src/country_info.rs:
```diff
@@ -0,0 +1,171 @@
+
+mod imp {
+    use super::*;
+
```
```suggestion
```

melix99 commented on src/country_list.rs:
```diff
@@ -0,0 +1,203 @@
+
+mod imp {
+    use super::*;
+
```
```suggestion
```

melix99 commented on src/country_info.rs:
```diff
@@ -0,0 +1,171 @@
+use gtk::glib;
+use gtk::prelude::*;
+use gtk::subclass::prelude::*;
+use once_cell::sync::Lazy;
```
This can be moved inside the `imp` mod block.

melix99 commented on src/country_list.rs:
```diff
@@ -0,0 +1,203 @@
+        fn item(&self, list_model: &Self::Type, position: u32) -> Option<glib::Object> {
+            let list = list_model.list();
+            if position as usize == list.len() {
+                // The last item in the list is always the invalid `CountryInfo`.
```
It's a bit weird to have an "invalid" item in the list model, this seems something that should be handled in the PhoneNumberInput widget instead, probably with an expression that can check if the selected item is invalid and show the "invalid" (or "unknown") text in case it is.

marhkb replied:
Yes agree. Not the best idea to let the user choose "invalid" from the list. I removed it. I have not found a way to make `CombRow` display custom text. So, now the subtitle of the row will have an error hint and nothing is selected in the combo box if the country code the user enters is invalid.

melix99 replied:
Ah, apparently the ComboRow::expression doesn't work if the selected item is null, so my suggestion doesn't work. So, I think your solution is ok.

marhkb commented on src/country_list.rs:
```diff
@@ -0,0 +1,213 @@
+        }
+
+        fn n_items(&self, list_model: &Self::Type) -> u32 {
+            list_model.list().len() as u32 + 1
```
This does not work. `self.list` is a `OnceCell`. So it should be `self.list.get().unwrap().len() as u32 + 1`. But I'm not sure whether this is preferable.

melix99 replied:
Oh, right, I didn't check that out.
